### ISSUE TITLE
Release size

### ISSUE THREAD

**Issue by cup** [BurntSushi/ripgrep#1481]
I currently use Silver Searcher

https://github.com/ggreer/the_silver_searcher

where the download is 181 KB:

https://github.com/k-takata/the_silver_searcher-win32/releases

I was surprised to find that the Windows RipGrep is 1.7 MB:

https://github.com/BurntSushi/ripgrep/releases

thats almost 10 times larger. Granted, even 1.7 MB is not large in the big
picture, but I am curious if someone has insight into the large difference.

**Comment by BurntSushi** [comment]
I'm not too familiar with Windows tooling around binary sizes, so perhaps you could do the analysis? But for Linux at least, the most obvious culprit is that the ripgrep releases statically link everything. As far as I can tell, your unofficial silver searcher release binaries do not, but I actually found it pretty hard to find the actual build commands in that repo, so I'm not sure.

I tried compiling ripgrep under a few different situations to show the differences in binary size, on Linux at least:

```
3.6M    ./rg-strip-linux-glibc-lto
4.3M    ./rg-strip-linux-glibc
4.5M    ./rg-strip-linux-musl
4.6M    ./rg-strip-linux-glibc-pcredy
5.1M    ./rg-strip-linux-glibc-pcrestatic
5.4M    ./rg-strip-linux-musl-pcrestatic
```

`glibc` means ripgrep was dynamically linked with the GNU's C standard library. `musl` means ripgrep was statically linked with MUSL (a small C standard library). `lto` means ripgrep was compiled with link time optimization (improves performance and decreases binary size but at the expense of significant compilation time). `pcredy` means ripgrep was dynamically linked with PCRE2 and `pcrestatic` means ripgrep was statically linked with PCRE2. For the most part, there should be no surprises here. When ripgrep is statically linked with MUSL and PCRE2, then its binary size is the biggest. This also just so happens to be the configuration for ripgrep's binary releases. The purpose here is to make the binaries as portable as possible. The cost is that the binary size is a bit bigger. I do this under the assumption that most people don't care about a couple of megabytes of disk space/bandwidth. For the small minority that do, they can compile their own build with more favorable settings. Or if that isn't good enough, they can use a different tool.

On Windows, ripgrep also statically links PCRE2, but I believe does not statically link the C standard library? So that right there, if it is indeed a difference, will be pretty sizable. You might consider compiling ripgrep without PCRE2 on Windows and compare the binary sizes.

Beyond that though, consider the following facts:

* The binary releases of ripgrep that I provide include two completely distinct regex engines. That is, Rust's regex engine and PCRE2. ripgrep can be compiled without PCRE2, which will decrease binary size, as shown above. On Linux at least, ripgrep can be dynamically linked with PCRE2, which will only modestly increase binary size.
* ripgrep has a fairly sophisticated glob matcher to make it fast and correct. ag uses `fnmatch` which is part of glibc.
* ripgrep also has two completely distinct recursive directory traversal implementations. One for single threaded use and one for parallel traversal. This provides substantial performance improvements over ag, which just uses standard `readdir` calls and interleaves its directory traversal with printing and searching. (It also uses actual recursion, which I assume will blow up on sufficiently nested directories.)
* ripgrep goes all in toward supporting gitignore correctly and with performance in mind. There is a substantial amount of code for this, including an implementation for matching many globs all at once with a single pass over a file path.
* ripgrep generally does a lot more work to make things go faster. It has an [entire library](https://github.com/BurntSushi/ripgrep/tree/master/grep-regex) dedicated to just analyzing regexes and specializing certain cases for speed. ag pretty much does none of this, other than detecting the trivial case of a non-regex literal pattern. ripgrep also puts a lot more effort to making printing fast, which does to some extent bloat code due to all of the function inlining. There's also more code in order to specialize certain cases in order to create fast paths. ag does none of this.
* ripgrep core is itself quite small. It's only about 4,000 lines of code, and the vast bulk of that (3,000 lines, which includes full docs for every flag) is just for defining and handling CLI arguments. In other words, ripgrep is split into many different crates. I didn't have to do this, but I did so that other tools in the ecosystem could benefit from my work. This has been a wild success with tools like tokei, fd, Cargo and even rustc itself incorporating some of these crates. The cost of course is de-coupling, which quite frequently leads to more code.
* ripgrep bundles an entire [encoding library](https://github.com/hsivonen/encoding_rs) for dealing with things like UTF-16 or any one of a number of other encodings. ag does none of this.

In summary, you should absolutely continue using the silver searcher if a couple MB of space on your hard drive is more important to you than performance, correctness and better Unicode support.

**Comment by cup** [comment]
@BurntSushi that is a fantastic response. I might look into ways to optimize the RipGrep size, but it seems RipGrep offers several benefits with regard to speed and features.

I hope you didnt take my post as a negative, in fact I think its great what youve posted as you can point anyone in the future here who has a similar question. If I manage to come up with some improvements I will post them here, thank you!

**Comment by cup** [comment]
I have looked at this more closely, and it seems the current releases are
already stripped. Further, RipGrep currently can disable features at runtime, 
but not compile time.

So the only path forward to shrinking the size would be to add compile time
options to disable some things. This would require first deciding what if
anything can be disabled.

I can understand if you are not interested in supporting that, but I wanted to
post my findings in case you wanted to comment on them, thanks again.

**Comment by BurntSushi** [comment]
Not really, no. Not unless it was trivial.

In general, I do not consider "I want to save a couple megabytes of disk space" to be a use case that is worth my time optimizing for. Sorry. If you have a real use case that literally demands this (not sure what that would be, maybe putting ripgrep on an embedded device?), then I'd recommend forking ripgrep and maintain a patch that cuts out everything you don't need.
